File: csrc/cpu/matching.cpp
// MASt3R Runtime - CPU Matching Implementation
// Copyright 2024 Delanoe Pirard / Aedelon. Apache 2.0.

#include <algorithm>
#include <chrono>
#include <cmath>
#include <numeric>
#include <vector>

#ifdef MAST3R_HAS_OPENMP
#include <omp.h>
#endif

#include "common/types.hpp"

namespace mast3r {
namespace cpu {

/**
 * L2 normalize descriptors in-place.
 */
void normalize_descriptors(float* desc, int N, int D) {
#ifdef MAST3R_HAS_OPENMP
    #pragma omp parallel for
#endif
    for (int i = 0; i < N; ++i) {
        float* row = desc + i * D;

        // Compute L2 norm
        float norm = 0.0f;
        for (int j = 0; j < D; ++j) {
            norm += row[j] * row[j];
        }
        norm = std::sqrt(norm);

        // Normalize
        if (norm > 1e-8f) {
            const float inv_norm = 1.0f / norm;
            for (int j = 0; j < D; ++j) {
                row[j] *= inv_norm;
            }
        }
    }
}

/**
 * Compute cosine similarity matrix.
 *
 * @param desc_1 [N1, D] normalized descriptors
 * @param desc_2 [N2, D] normalized descriptors
 * @param sim [N1, N2] output similarity matrix
 */
void compute_similarity(
    const float* desc_1, const float* desc_2,
    float* sim,
    int N1, int N2, int D
) {
#ifdef MAST3R_HAS_OPENMP
    #pragma omp parallel for collapse(2)
#endif
    for (int i = 0; i < N1; ++i) {
        for (int j = 0; j < N2; ++j) {
            float dot = 0.0f;
            for (int k = 0; k < D; ++k) {
                dot += desc_1[i * D + k] * desc_2[j * D + k];
            }
            sim[i * N2 + j] = dot;
        }
    }
}

/**
 * Reciprocal matching.
 *
 * @param desc_1 [H*W, D] descriptors from view 1
 * @param desc_2 [H*W, D] descriptors from view 2
 * @param conf_1 [H*W] confidence (optional, can be nullptr)
 * @param conf_2 [H*W] confidence (optional, can be nullptr)
 * @param height image height
 * @param width image width
 * @param desc_dim descriptor dimension
 * @param config matching configuration
 * @return MatchResult with correspondences
 */
MatchResult reciprocal_match(
    const float* desc_1, const float* desc_2,
    const float* conf_1, const float* conf_2,
    int height, int width, int desc_dim,
    const MatchingConfig& config
) {
    using Clock = std::chrono::high_resolution_clock;
    auto t0 = Clock::now();

    const int N = height * width;

    // Copy and normalize descriptors
    std::vector<float> desc_1_norm(desc_1, desc_1 + N * desc_dim);
    std::vector<float> desc_2_norm(desc_2, desc_2 + N * desc_dim);
    normalize_descriptors(desc_1_norm.data(), N, desc_dim);
    normalize_descriptors(desc_2_norm.data(), N, desc_dim);

    // Compute similarity matrix [N, N]
    std::vector<float> sim(N * N);
    compute_similarity(
        desc_1_norm.data(), desc_2_norm.data(),
        sim.data(), N, N, desc_dim
    );

    // Apply confidence weighting if provided
    if (conf_1 != nullptr && conf_2 != nullptr) {
#ifdef MAST3R_HAS_OPENMP
        #pragma omp parallel for collapse(2)
#endif
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                sim[i * N + j] *= conf_1[i] * conf_2[j];
            }
        }
    }

    // Find nearest neighbors
    std::vector<int> nn_12(N);  // Forward: desc_1 -> desc_2
    std::vector<int> nn_21(N);  // Backward: desc_2 -> desc_1

#ifdef MAST3R_HAS_OPENMP
    #pragma omp parallel for
#endif
    for (int i = 0; i < N; ++i) {
        // Forward NN
        int best_j = 0;
        float best_sim = sim[i * N];
        for (int j = 1; j < N; ++j) {
            if (sim[i * N + j] > best_sim) {
                best_sim = sim[i * N + j];
                best_j = j;
            }
        }
        nn_12[i] = best_j;
    }

#ifdef MAST3R_HAS_OPENMP
    #pragma omp parallel for
#endif
    for (int j = 0; j < N; ++j) {
        // Backward NN
        int best_i = 0;
        float best_sim = sim[j];
        for (int i = 1; i < N; ++i) {
            if (sim[i * N + j] > best_sim) {
                best_sim = sim[i * N + j];
                best_i = i;
            }
        }
        nn_21[j] = best_i;
    }

    // Collect reciprocal matches
    std::vector<int64_t> idx_1, idx_2;
    std::vector<float> confidence;

    if (config.reciprocal) {
        for (int i = 0; i < N; ++i) {
            const int j = nn_12[i];
            if (nn_21[j] == i) {
                const float score = sim[i * N + j];
                if (score >= config.confidence_threshold) {
                    idx_1.push_back(i);
                    idx_2.push_back(j);
                    confidence.push_back(score);
                }
            }
        }
    } else {
        for (int i = 0; i < N; ++i) {
            const int j = nn_12[i];
            const float score = sim[i * N + j];
            if (score >= config.confidence_threshold) {
                idx_1.push_back(i);
                idx_2.push_back(j);
                confidence.push_back(score);
            }
        }
    }

    // Keep top-K by confidence
    const int num_matches = static_cast<int>(idx_1.size());
    if (num_matches > config.top_k) {
        // Sort indices by confidence (descending)
        std::vector<int> order(num_matches);
        std::iota(order.begin(), order.end(), 0);
        std::partial_sort(
            order.begin(), order.begin() + config.top_k, order.end(),
            [&confidence](int a, int b) {
                return confidence[a] > confidence[b];
            }
        );

        // Keep top-K
        std::vector<int64_t> new_idx_1(config.top_k);
        std::vector<int64_t> new_idx_2(config.top_k);
        std::vector<float> new_conf(config.top_k);
        for (int k = 0; k < config.top_k; ++k) {
            new_idx_1[k] = idx_1[order[k]];
            new_idx_2[k] = idx_2[order[k]];
            new_conf[k] = confidence[order[k]];
        }
        idx_1 = std::move(new_idx_1);
        idx_2 = std::move(new_idx_2);
        confidence = std::move(new_conf);
    }

    // Convert flat indices to 2D coordinates
    const int final_count = static_cast<int>(idx_1.size());
    std::vector<float> pts2d_1(final_count * 2);
    std::vector<float> pts2d_2(final_count * 2);

    for (int k = 0; k < final_count; ++k) {
        pts2d_1[k * 2 + 0] = static_cast<float>(idx_1[k] % width);  // x
        pts2d_1[k * 2 + 1] = static_cast<float>(idx_1[k] / width);  // y
        pts2d_2[k * 2 + 0] = static_cast<float>(idx_2[k] % width);
        pts2d_2[k * 2 + 1] = static_cast<float>(idx_2[k] / width);
    }

    auto t1 = Clock::now();

    MatchResult result;
    result.idx_1 = std::move(idx_1);
    result.idx_2 = std::move(idx_2);
    result.pts2d_1 = std::move(pts2d_1);
    result.pts2d_2 = std::move(pts2d_2);
    result.pts3d_1.resize(final_count * 3, 0.0f);  // Placeholder
    result.pts3d_2.resize(final_count * 3, 0.0f);
    result.confidence = std::move(confidence);
    result.match_ms = std::chrono::duration<float, std::milli>(t1 - t0).count();

    return result;
}

}  // namespace cpu
}  // namespace mast3r

File: mast3r_runtime/backends/matching.py
"""Descriptor matching utilities for MASt3R runtime.

Pure numpy implementation for lightweight deployment.

Copyright 2024 Delanoe Pirard / Aedelon. Apache 2.0.
"""

from __future__ import annotations

import time

import numpy as np
from numpy.typing import NDArray

from ..core.engine_interface import MatchResult


def normalize_descriptors(desc: NDArray[np.float32]) -> NDArray[np.float32]:
    """L2 normalize descriptors.

    Args:
        desc: Descriptors [H, W, D] or [N, D]

    Returns:
        Normalized descriptors with same shape
    """
    norm = np.linalg.norm(desc, axis=-1, keepdims=True)
    norm = np.maximum(norm, 1e-8)  # Avoid division by zero
    return desc / norm


def compute_similarity_matrix(
    desc_1: NDArray[np.float32],
    desc_2: NDArray[np.float32],
) -> NDArray[np.float32]:
    """Compute cosine similarity between all descriptor pairs.

    Args:
        desc_1: Descriptors [N1, D]
        desc_2: Descriptors [N2, D]

    Returns:
        Similarity matrix [N1, N2]
    """
    # Normalize
    desc_1_norm = normalize_descriptors(desc_1)
    desc_2_norm = normalize_descriptors(desc_2)

    # Cosine similarity via dot product
    return desc_1_norm @ desc_2_norm.T


def reciprocal_match(
    desc_1: NDArray[np.float32],
    desc_2: NDArray[np.float32],
    conf_1: NDArray[np.float32] | None = None,
    conf_2: NDArray[np.float32] | None = None,
    pts3d_1: NDArray[np.float32] | None = None,
    pts3d_2: NDArray[np.float32] | None = None,
    top_k: int = 512,
    reciprocal: bool = True,
    confidence_threshold: float = 0.5,
) -> MatchResult:
    """Reciprocal top-K matching between descriptors.

    Finds mutual nearest neighbors between two sets of descriptors.

    Args:
        desc_1: Descriptors from view 1 [H, W, D]
        desc_2: Descriptors from view 2 [H, W, D]
        conf_1: Optional confidence from view 1 [H, W]
        conf_2: Optional confidence from view 2 [H, W]
        pts3d_1: Optional 3D points from view 1 [H, W, 3]
        pts3d_2: Optional 3D points from view 2 [H, W, 3]
        top_k: Maximum number of matches to return
        reciprocal: Whether to use reciprocal matching
        confidence_threshold: Minimum match confidence

    Returns:
        MatchResult with correspondences
    """
    timing = {}
    t0 = time.perf_counter()

    _H, W, D = desc_1.shape

    # Flatten spatial dimensions
    desc_1_flat = desc_1.reshape(-1, D)  # [N, D] where N = H * W
    desc_2_flat = desc_2.reshape(-1, D)

    # Compute similarity matrix
    sim = compute_similarity_matrix(desc_1_flat, desc_2_flat)  # [N, N]

    # Apply confidence weighting if provided
    if conf_1 is not None and conf_2 is not None:
        conf_1_flat = conf_1.reshape(-1)  # [N]
        conf_2_flat = conf_2.reshape(-1)  # [N]
        conf_weight = np.outer(conf_1_flat, conf_2_flat)  # [N, N]
        sim = sim * conf_weight

    if reciprocal:
        # Find mutual nearest neighbors
        # Forward: for each in desc_1, find best match in desc_2
        nn_12 = np.argmax(sim, axis=1)  # [N]
        # Backward: for each in desc_2, find best match in desc_1
        nn_21 = np.argmax(sim, axis=0)  # [N]

        # Reciprocal check: nn_21[nn_12[i]] == i
        idx_1_all = np.arange(len(nn_12))
        reciprocal_mask = nn_21[nn_12] == idx_1_all

        # Get reciprocal matches
        idx_1 = idx_1_all[reciprocal_mask]
        idx_2 = nn_12[reciprocal_mask]

        # Get match confidences
        confidence = sim[idx_1, idx_2]

    else:
        # Simple top-K matching without reciprocity
        # For each point in desc_1, find best match in desc_2
        nn_12 = np.argmax(sim, axis=1)  # [N]
        idx_1 = np.arange(len(nn_12))
        idx_2 = nn_12
        confidence = sim[idx_1, idx_2]

    # Filter by confidence threshold
    valid_mask = confidence >= confidence_threshold
    idx_1 = idx_1[valid_mask]
    idx_2 = idx_2[valid_mask]
    confidence = confidence[valid_mask]

    # Keep top-K by confidence
    if len(idx_1) > top_k:
        top_indices = np.argsort(confidence)[-top_k:]
        idx_1 = idx_1[top_indices]
        idx_2 = idx_2[top_indices]
        confidence = confidence[top_indices]

    # Convert flat indices to 2D coordinates
    pts2d_1 = np.stack([idx_1 % W, idx_1 // W], axis=1).astype(np.float32)  # [N, 2] (x, y)
    pts2d_2 = np.stack([idx_2 % W, idx_2 // W], axis=1).astype(np.float32)

    # Get 3D points at match locations
    if pts3d_1 is not None and pts3d_2 is not None:
        pts3d_1_flat = pts3d_1.reshape(-1, 3)
        pts3d_2_flat = pts3d_2.reshape(-1, 3)
        matched_pts3d_1 = pts3d_1_flat[idx_1]
        matched_pts3d_2 = pts3d_2_flat[idx_2]
    else:
        matched_pts3d_1 = np.zeros((len(idx_1), 3), dtype=np.float32)
        matched_pts3d_2 = np.zeros((len(idx_2), 3), dtype=np.float32)

    timing["match_ms"] = (time.perf_counter() - t0) * 1000

    return MatchResult(
        idx_1=idx_1.astype(np.int64),
        idx_2=idx_2.astype(np.int64),
        pts2d_1=pts2d_1,
        pts2d_2=pts2d_2,
        pts3d_1=matched_pts3d_1,
        pts3d_2=matched_pts3d_2,
        confidence=confidence.astype(np.float32),
        timing_ms=timing,
    )


def spatial_filter_matches(
    match_result: MatchResult,
    max_displacement: float = 100.0,
) -> MatchResult:
    """Filter matches by spatial consistency.

    Removes matches where the 2D displacement is too large.

    Args:
        match_result: Input matches
        max_displacement: Maximum allowed pixel displacement

    Returns:
        Filtered MatchResult
    """
    # Compute 2D displacement
    displacement = np.linalg.norm(match_result.pts2d_1 - match_result.pts2d_2, axis=1)
    valid_mask = displacement <= max_displacement

    return MatchResult(
        idx_1=match_result.idx_1[valid_mask],
        idx_2=match_result.idx_2[valid_mask],
        pts2d_1=match_result.pts2d_1[valid_mask],
        pts2d_2=match_result.pts2d_2[valid_mask],
        pts3d_1=match_result.pts3d_1[valid_mask],
        pts3d_2=match_result.pts3d_2[valid_mask],
        confidence=match_result.confidence[valid_mask],
        timing_ms=match_result.timing_ms.copy(),
    )
